Add Python installation guide and Cotainr documentation (#127)
Lumi-supercomputer/lumi-userguide: docs/software/installing/container-wrapper.md

[conda]: https://docs.conda.io/en/latest/
[conda-env]: https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#sharing-an-environment
[mamba]: https://github.com/mamba-org/mamba
[pip]: https://pip.pypa.io/en/latest/
[pip-req]: https://pip.pypa.io/en/latest/reference/requirements-file-format/
[tykky-github]: https://github.com/CSCfi/hpc-container-wrapper

[python-install]: ./python.md
[softwarestacks]: ../../runjobs/lumi_env/softwarestacks.md

# Tykky installation wrapper

> The [Tykky][tykky-github] installation wrapper (formerly the "LUMI container

Reviewer: ```suggestion
The [Tykky][tykky-github] installation wrapper (formerly called the "LUMI container
```

lessen the number of files on large parallel file systems.

Additionally, the Tykky installation wrapper will generate wrappers so that
installed software can be used (almost) as if it was not containerized.
Depending on tool selection and settings, either the whole host file system or
a limited subset is visible during execution and installation. This means that
it is possible to wrap installations using e.g. mpi4py while still relying on
the host provided MPI installation.

Tykky is a general purpose installation wrapper that supports wrapping:

- **Existing installations on the filesystem**: Mainly to reduce the I/O load
  and improve startup times, but may also be used to containerize existing
  installations that cannot be re-installed.
- **Existing Singularity/Apptainer containers**: Mainly to hide the need for
  using the container runtime from the user.
- **Conda installations**: Directly wrap a [Conda][conda] installation based on
  a [Conda environment file][conda-env].
- **Pip installations**: Directly wrap a [pip][pip] installation based on a
  [requirements.txt file][pip-req].

!!! warning "Tykky is NOT generally recommended for managing Conda/pip installations"

>     For some use cases, Tykky is an excellent tools for managing Conda/pip

Reviewer: ```suggestion
    For some use cases, Tykky is an excellent tool for managing Conda/pip
```

    installations. For other use cases, there are better alternatives. See the
    [installing Python packages][python-install] guide for an overview the
    recommended ways to manage Python installations, including Conda/pip
    installations.

!!! warning "Tykky has some limitations"
    Please be aware of the [limitations when using Tykky
    ](https://github.com/CSCfi/hpc-container-wrapper#limitations) before 
    wrapping your installations.

!!! warning "Tykky is experimental software"
    As the Tykky installation wrapper is still under development, some of the
    more advanced features might change in exact usage and API.

## Examples of using Tykky

The tools provided by Tykky are accessible by loading the
`lumi-container-wrapper` module that is available in the `LUMI` and `CrayEnv`
[software stacks][softwarestacks].

```bash
$ module load LUMI
$ module load lumi-container-wrapper
```

Once the module has been loaded, you can use one of the front-end tools
`conda-containerize`, `pip-containerize`, `wrap-container`, and `wrap-install`.

### Wrapping a basic Conda installation

> In order to wrap a basic [Conda][conda] installation, create an `<install_dir>`

Reviewer: ```suggestion
In order to wrap a basic [Conda][conda] installation, create an installation directory
```
